compiler/executable/pipeline.rs

```diff
@@ -49,7 +51,7 @@ pub enum ExecutableStage {
     Offset(Arc<OffsetExecutable>),
     Limit(Arc<LimitExecutable>),
     Require(Arc<RequireExecutable>),
-    Reduce(Arc<ReduceExecutable>),
+    Reduce(Arc<ReduceStageExecutable>),
```

Comment:
`ReduceExecutable`, used for return operations as well, doesn't need to contain output positions mappings. That's why we create a new `ReduceStageExecutable`, that contains `ReduceExecutable` and these output mappings.

Reply:
@dmitrii-ubskii With `ExecutableStage::Reduce(Arc<ReduceExecutor>)`?

Reply:
No. `FooExecutable`s are stages, and contain `BarExecutor`s.

Reply:
Okay... So my current
```
#[derive(Debug, Clone)]
pub struct ReduceStageExecutable {
    pub reduce_executable: Arc<ReduceExecutable>,
    pub output_row_mapping: HashMap<Variable, VariablePosition>, // output_row = (group_vars, reduce_outputs)
}

#[derive(Debug, Clone)]
pub struct ReduceExecutable {
    pub reductions: Vec<ReduceInstruction<VariablePosition>>,
    pub input_group_positions: Vec<VariablePosition>,
}
```
to be changed to
```
#[derive(Debug, Clone)]
pub struct ReduceExecutable {
    pub executor: Arc<ReduceExecutor>,
    pub output_row_mapping: HashMap<Variable, VariablePosition>, // output_row = (group_vars, reduce_outputs)
}

#[derive(Debug, Clone)]
pub struct ReduceExecutor {
    pub reductions: Vec<ReduceInstruction<VariablePosition>>,
    pub input_group_positions: Vec<VariablePosition>,
}
```
@flyingsilverfin does it make sense to you?

Reply:
Err an `Executable` shouldn't have a reference to an `Executor` - the other way around however could be true.

Reply:
Eg: 
```
pub struct FetchStageExecutor<Snapshot: ReadableSnapshot> {
    executable: Arc<ExecutableFetch>,
    _phantom: PhantomData<Snapshot>,
}
```

Reply:
Our naming convention is `TypeQL -> Translated -> Annotated -> Executable -> StageExecutor(executable)`. The main case where we have an extra layer is for `match`: `MatchExecutable -> MatchStageExecutable(MatchExecutor(MatchExecutable))`, although the extra `MatchExecutor` is created within the iterator for the `MatchStageExecutable`

Reply:
I'm kinda lost as I feel like some of the statements are contradicting. 

With my current naming, we have:

`executor/pipeline/reduce`:
```
pub struct ReduceStageExecutor<PreviousStage> {
    executable: Arc<ReduceStageExecutable>,
    previous: PreviousStage,
}
```

`compiler/executable/reduce`:
```
pub struct ReduceStageExecutable {
    pub reduce_executable: Arc<ReduceExecutable>,
    pub output_row_mapping: HashMap<Variable, VariablePosition>, // output_row = (group_vars, reduce_outputs)
}

pub struct ReduceExecutable {
    pub reductions: Vec<ReduceInstruction<VariablePosition>>,
    pub input_group_positions: Vec<VariablePosition>,
}
```

What's the correct naming, then?

Reply:
Why do you have separate `ReduceStageExecutable` and `ReduceExecutable`? We don't have a **StageExecutable** for any other tyoe pf stage? We only ever have a `XXXStageExcutor` that has an `XXXExecutable` inside?

Reply:
@flyingsilverfin I've explained it in the first comment of this thread. I could send `output_row_mapping` as an additional argument everywhere, but it doesn't really make sense, and more comfortable to wrap it together with the executable.
Specifically, I had a case in `compile_return_operation` in `compiler`, where I didn't have and didn't need `output_row_mapping` as we are working with the return statement already, where the mapping is direct and automatic. So we made this refactoring with @krishnangovindraj